### ISSUE TITLE
Add reward shaping and policy shaping to DQN

### ISSUE THREAD

**Issue by zhujl1991** [thu-ml/tianshou#279]
Do we have any plans to add [reward shaping and policy shaping](https://www.aclweb.org/anthology/2020.acl-main.566.pdf) to DQN? Looks like the reward shaping requires overwriting [here](https://github.com/thu-ml/tianshou/blob/master/tianshou/policy/modelfree/dqn.py#L93), and policy shaping requires overwriting [here](https://github.com/thu-ml/tianshou/blob/master/tianshou/policy/modelfree/dqn.py#L106).

**Comment by Trinkle23897** [comment]
You're right, and also reward shaping can somehow write in `gym.Wrapper` or `collector.preprocess_fn` (if treating F_D as a part of the environment).
Currently we don't have any plan. But you are welcome to make a PR to let them compatible with *all* policies.

**Comment by zhujl1991** [comment]
> You're right, and also reward shaping can somehow write in `gym.Wrapper` or `collector.preprocess_fn` (if treating F_D as a part of the environment).
> Currently we don't have any plan. But you are welcome to make a PR to let them compatible with _all_ policies.

AFAIU, the shaping is only applied to DQN. What do you mean by `compatible with *all* policies.`?

**Comment by Trinkle23897** [comment]
Because in section `3.3 Policy Shaping` the authors doesn't claim that this algorithm is only compatible with DQN. My understanding is that this policy shaping method can be adapted to any off-policy + Q-value function approximation algorithms.

But even if the policy shaping can only be adapted to DQN, I don't think it is hard to extend to DQN families -- C51, QRDQN, etc.
